builtin/credential/ldap/backend.go

```diff
@@ -212,70 +251,89 @@ func getUserDN(cfg *ConfigEntry, c *ldap.Conn, bindDN string) (string, error) {
 	return userDN, nil
 }
 
+/*
+ * getLdapGroups queries LDAP and returns a slice describing the set of groups the authenticated user is a member of.
+ *
+ * The search query is constructed according to cfg.GroupFilter, and run in context of cfg.GroupDN.
+ * Groups will be resolved from the query results by following the attribute defined in cfg.GroupAttr.
+ *
+ * cfg.GroupFilter is a go template and is compiled with the following context: [UserDN, Username]
+ *    UserDN - The DN of the authenticated user
+ *    UserName - The Username of the authenticated user
+ *
+ * Example:
+ *   cfg.GroupFilter = "(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={{.UserDN}}))"
+ *   cfg.GroupDN     = "OU=Groups,DC=myorg,DC=com"
+ *   cfg.GroupAttr   = "cn"
+ *
+ * NOTE - If cfg.GroupFilter is empty, no query is performed and an empty result slice is returned.
+ *
+ */
 func getLdapGroups(cfg *ConfigEntry, c *ldap.Conn, userDN string, username string) ([]string, error) {
 	// retrieve the groups in a string/bool map as a structure to avoid duplicates inside
 	ldapMap := make(map[string]bool)
-	// Fetch the optional memberOf property values on the user object
-	// This is the most common method used in Active Directory setup to retrieve the groups
+
+	if cfg.GroupFilter == "" {
+		cfg.Logger.Printf("[WARN] GroupFilter is empty, cannot query groups for user", cfg.GroupFilter)
+		return make([]string, 0), nil
+	}
+
+	// If groupfilter was defined, resolve it as a Go template and use the query for
+	// returning the user's groups
+	cfg.Logger.Printf("[DEBUG] Compiling group filter %s", cfg.GroupFilter)
+
+	// Parse the configuration as a template.
+	// Example template "(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={{.UserDN}}))"
+	t, err := template.New("queryTemplate").Parse(cfg.GroupFilter)
+	if err != nil {
+		return nil, fmt.Errorf("LDAP search failed due to template compilation error: %v", err)
+	}
+
+	// Build context to pass to template - we will be exposing userDN and username.
+	context := struct {
+		UserDN   string
+		Username string
+	}{
+		ldap.EscapeFilter(userDN),
+		ldap.EscapeFilter(username),
+	}
+
+	var renderedQuery bytes.Buffer
+	t.Execute(&renderedQuery, context)
+	cfg.Logger.Printf("[DEBUG] LDAP search with query: %s", renderedQuery.String())
+
+	// TODO - ok to proceed if renderedQuery length is 0?
 	result, err := c.Search(&ldap.SearchRequest{
-		BaseDN: userDN,
```

Comment:
That was the logic that was added in #1245 and was biased towards a specific configuration. In fact, with #1245 the result was an additional query being run against the LDAP server that many users did not need and could not turn off.

It is no longer the default case - if you want to resolve the group relationship through the user object rather than through group objects, you need to say so in the configuration:

```
groupfilter: "(&(objectClass=person)(uid={{.Username}}))",
groupattr: "memberOf",
groupdn: "<search base for my org's user objects"
```

The code is simpler, the configuration is easier to reason about, but there is a slight change in behavior in this regard with the `memberOf` change introduced in #1245. This should be called out in the release notes.